### PR TITLE
Expose fuzzy search for issues/pulls

### LINKED ISSUE #29685: Expose fuzzy search for issues/pulls
Currently the backend option to use fuzzy search for issues is not exposed in the WebUI.

https://github.com/go-gitea/gitea/blob/7fdc0481538151d8a5ed3ec2a32639950f5d8ac6/modules/indexer/issues/indexer.go#L283

Is the backend func in question witch si controled via **SearchOptions**

To enable fuzzy search set
https://github.com/go-gitea/gitea/blob/7fdc0481538151d8a5ed3ec2a32639950f5d8ac6/modules/indexer/issues/internal/model.go#L77
to `true`.

### PR DESCRIPTION
**Pull request by 6543**
close  #29685

### CODE REVIEW

**Comment by 6543** [comment]
TODO: rebase ontop of https://github.com/go-gitea/gitea/pull/29530

File: modules/indexer/issues/meilisearch/meilisearch.go [hunk 1]
```diff
@@ -234,6 +234,7 @@ func (b *Indexer) Search(ctx context.Context, options *internal.SearchOptions) (
 	}
 
 	return &internal.SearchResult{
+		// TODO: if nonFuzzyWorkaround is not needed anymore this should deliver correct counts again
```
Comment (6543) [inline modules/indexer/issues/meilisearch/meilisearch.go]: unrelated nit ... but good to have it documented

once merged, I'll create an issue so others do not open new one and we have it tracked e.g. will link it to upstream too ...

Comment (6543) [reply modules/indexer/issues/meilisearch/meilisearch.go]: #29740 fix it :)

**Comment by silverwind** [comment]
Screenshots?

**Comment by 6543** [comment]
I looks like other fuzzy search boxes ... but i can provide some

**Comment by silverwind** [comment]
Oh, and fix the conflict first please.

**Comment by 6543** [comment]
image
image

**Comment by silverwind** [comment]
I think we want to eventually introduce github-like syntax to issue search. I guess this box is fine for now, but I would prefer it to be moved to a option in the syntax, e.g. `mode=fuzzy`.

BTW, "match" mode is a confusing term for me. What does it do? Shouldn't it be named like "Exact"?

**Comment by 6543** [comment]
yes i want to have it also more flexible ... but that will currently also improve things

**Comment by silverwind** [comment]
I recalling reading that the issue/pr search already supports some kind of `key:value` syntax. Could you check the code if this is true? If so, I would the `mode` option there add it there instead (feel free to use another name).

**Comment by 6543** [comment]
I would like to see it more like gitlab ... as i dont want to expect each normal user to remember all filter keywords ...

**Comment by silverwind** [comment]
Imho, keywords are the way to go. We can't bloat the UI forever with adding buttons. If GitHub would do that, there would bei 15+ additional buttons by the search.

**Comment by 6543** [comment]
I agree so that's why i suggest to do it similar like gitlab

**Comment by silverwind** [comment]
I guess I should check how GitLab does it, not familiar with it.

**Comment by 6543** [comment]
> I recalling reading somewhere that the issue/pr search already supports some kind of `key:value` syntax. Could you check the code if this is true? If so, we should add the `mode` option there instead (feel free to use another name).

Could you point me to that part?

**Comment by silverwind** [comment]
It was in one of the recent PRs, let's see if I find it.

**Comment by silverwind** [comment]
It was https://github.com/go-gitea/gitea/pull/29429/files which uses such ":" queries against the API for finding commits.

**Comment by 6543** [comment]
are you sure - the frontend code might introduce that filters but I am not aware of them to exist in the backend at all !!!

@silverwind I might check that within the codebase later to be sure ...

**Comment by silverwind** [comment]
Haven't checked if they actually exits. For the record here is the line from that PR, disregard the incorrect query escaping:

```js
`${this.repoLink}/commits/branch/${this.repoBranch}/search?q=after:${min}, before:${max}, author:`
```

**Comment by sahinakkaya** [comment]
Yeah, the backend code must exist because my PR works with its current state. I can't share the link to code because I am on phone right now but it shouldn't be hard to find. I might share the link later.

**Comment by denyskon** [comment]
The tooltip of issue search button says that it exists too, spotted in my search boxes PR : laughing:

**Comment by sahinakkaya** [comment]
Here is the relevant function from backend:

https://github.com/go-gitea/gitea/blob/43aa914b1750add631698a7fd57d76f743b0489b/routers/web/repo/commit.go#L184-L211

**Comment by 6543** [comment]
Uh it was added at https://github.com/go-gitea/gitea/pull/6479

And just supportbmy argument: even I did not knowbit existed: no docu and no UI that would explain it -> bad UX

**Comment by 6543** [comment]
PS: single tooltip is not enough!

**Comment by 6543** [comment]
@silverwind -> [Video of gitlab fitler concept](https://cloud.obermui.de/s/AFRfXRP7ygb5FYf)

**Comment by 6543** [comment]
@silverwind I personally would say we should go with this current solution and work on mentioned generic filter in a dedicated pull ... but to be honest I dont think this will be me who would create that followup.

anyway if you think we should do it right the first time I'll create an issue instead and describe it in more details ... and we can close the discussion here ...

**Comment by silverwind** [comment]
> @silverwind -> [Video of gitlab fitler concept](https://cloud.obermui.de/s/AFRfXRP7ygb5FYf)

Looks very similar to what Kibana has, but slightly worst 😆. Still better than GitHub were one has no  autocomplete at all.

**Comment by silverwind** [comment]
Tested it:

1. Box is missing in repo issue/pulls list
2. The addition of the box causes problems on certain screen widths:

image

3. I really have to question the term "Match". I think "Exact" would be much better.

**Comment by 6543** [comment]
>     3. I really have to question the term "Match". I think "Exact" would be much better.

yes that is better ... we did touch the locales not long ago  ... so i guess it is fine to change it ...

**Comment by silverwind** [comment]
Yeah, let's change it and rename the translation key as well.

**Review by silverwind — APPROVED** [review]

**Comment by silverwind** [comment]
Oh, forgot to ask: Does this change add the box on repo issue/pr lists too now? I think it should.

**Comment by 6543** [comment]
No i can add that

**Comment by silverwind** [comment]
I'd say do it. We need to stay consistent. Ideally via shared template.

**Comment by silverwind** [comment]
One idea I had is we could determine fuzzy vs. exact by presence of quotes around the term, e.g. `'term'` and `"term"` would search the term exact. Could later even be used to combine fuzzy and non-fuzzy with `term1 "term2"`.

**Comment by 6543** [comment]
Yes but two thungs against that idea:
- most indexers (the backemd(s) we depend on) have an option to tuggle fuzzines (meilisearch would be the exeption and do support it)
- it would be a breaking change as now fizzines is disabled by default then it is enabled by default
- we need an UX concept that explains that feature to the users

So blocking is the first one ... second is just anojing users

**Comment by denyskon** [comment]
> I'd say do it. We need to stay consistent. Ideally via shared template. I think there is one already with search box + dropdown.

Yes, there should be a "shared/search/combo_fuzzy"

**Comment by silverwind** [comment]
> we need an UX concept that explains that feature to the users

I disagree. Google also does not explain their query syntax on the UI. It's just something people learn over time.

**Comment by 6543** [comment]
Well i think we can agree that we disagree in this regards, but that should not hold back the current state just the future paning for the search boxes ... ps: i want more opinions this regards anyway and will go with whatever most can work best with

**Comment by 6543** [comment]
@silverwind I try to enable it for repo-issue-search too ... but just discovered we have a completly different search infrastructure for that :face_with_spiral_eyes: 

I'll create an issue but I think this NEEDS a refactor :)

**Comment by silverwind** [comment]
Oh, I thought this was just the matter of adding the dropdown. I guess the backend side must be really messy 😆.

**Comment by 6543** [comment]
@denyskon wana be secomd reviewer?

**Comment by denyskon** [comment]
I'll check, but not today 😅

**Review by lunny — APPROVED** [review]

File: templates/user/dashboard/issues.tmpl [hunk 2]
```diff
@@ -51,9 +51,11 @@
 							<input type="hidden" name="type" value="{{$.ViewType}}">
 							<input type="hidden" name="sort" value="{{$.SortType}}">
 							<input type="hidden" name="state" value="{{$.State}}">
-							{{template "shared/search/input" dict "Value" $.Keyword}}
-							<button id="issue-list-quick-goto" class="ui small icon button tw-hidden" data-tooltip-content="{{ctx.Locale.Tr "explore.go_to"}}">{{svg "octicon-hash"}}</button>
-							{{template "shared/search/button"}}
+							{{if .PageIsPulls}}
+								{{template "shared/search/combo_fuzzy" dict "Value" $.Keyword "IsFuzzy" $.IsFuzzy "Placeholder" (ctx.Locale.Tr "search.pull_kind") "Tooltip" (ctx.Locale.Tr "explorer.go")}}
+							{{else}}
+								{{template "shared/search/combo_fuzzy" dict "Value" $.Keyword "IsFuzzy" $.IsFuzzy "Placeholder" (ctx.Locale.Tr "search.issue_kind") "Tooltip" (ctx.Locale.Tr "explorer.go")}}
```
Comment (kerwin612) [inline templates/user/dashboard/issues.tmpl:57]: Not sure if it was a mistake, or if the new localized `explorer.go` forgot to commit? There is no `explorer.go` in the main branch, only `explorer.go_to`.

Comment (6543) [reply templates/user/dashboard/issues.tmpl:57]: -> #31348

**Pull request merged**